hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/upgrade/TestDNDataDistributionFinalization.java

```diff
@@ -0,0 +1,337 @@
+   * This test specifically checks the conditions in populatePendingDeletionMetadata:
+   * 1. Pre-finalization: handlePreDataDistributionFeature path
+   * 2. Post-finalization: handlePostDataDistributionFeature path
+   * 3. Missing metadata: getAggregatePendingDelete path
+   */
+  @Test
+  public void testDataDistributionUpgradeScenario() throws Exception {
+    init(new OzoneConfiguration(), null, false);
+
+    // Verify initial state - DATA_DISTRIBUTION should not be finalized yet
+    assertEquals(HDDSLayoutFeature.HBASE_SUPPORT.layoutVersion(),
+        scmClient.getScmInfo().getMetaDataLayoutVersion());
+
+    // Create some data and delete operations to trigger pending deletion logic
+    String keyName1 = "testKey1";
+    String keyName2 = "testKey2";
+    byte[] data = new byte[1024];
+
+    // Write some keys
+    try (OzoneOutputStream out = bucket.createKey(keyName1, data.length)) {
+      out.write(data);
+    }
+    try (OzoneOutputStream out = bucket.createKey(keyName2, data.length)) {
+      out.write(data);
+    }
+
+    // Delete one key to create pending deletion blocks
+    bucket.deleteKey(keyName1);
+
+    // Validate pre-finalization state
+    validatePreDataDistributionFeatureState();
+
+    // Now trigger finalization
+    finalizationFuture = Executors.newSingleThreadExecutor().submit(
+        () -> {
+          try {
+            scmClient.finalizeScmUpgrade(CLIENT_ID);
+          } catch (IOException ex) {
+            LOG.info("finalization client failed. This may be expected if the" +
+                " test injected failures.", ex);
+          }
+        });
+
+    // Wait for finalization to complete
+    finalizationFuture.get();
+    TestHddsUpgradeUtils.waitForFinalizationFromClient(scmClient, CLIENT_ID);
+
+    // Verify finalization completed
+    assertEquals(HDDSLayoutFeature.DATA_DISTRIBUTION.layoutVersion(),
+        scmClient.getScmInfo().getMetaDataLayoutVersion());
+
+    // Create more data and deletions to test post-finalization behavior
+    String keyName3 = "testKey3";
+    try (OzoneOutputStream out = bucket.createKey(keyName3, data.length)) {
+      out.write(data);
+    }
+    bucket.deleteKey(keyName2);
+    bucket.deleteKey(keyName3);
+
+    // Validate post-finalization state
+    validatePostDataDistributionFeatureState();
+  }
+
+  /**
+   * Test specifically for the missing metadata scenario that triggers
+   * the getAggregatePendingDelete code path.
+   */
+  @Test
+  public void testMissingPendingDeleteMetadataRecalculation() throws Exception {
+    init(new OzoneConfiguration(), null, true);
+
+    // Create and delete keys to generate some pending deletion data
+    String keyName = "testKeyForRecalc";
+    byte[] data = new byte[2048];
+
+    try (OzoneOutputStream out = bucket.createKey(keyName, data.length)) {
+      out.write(data);
+    }
+    bucket.deleteKey(keyName);
+
+    // Wait for finalization
+    finalizationFuture.get();
```

Comment:
Can we move the file deletion before start the finalization process, to make sure the recalculation to happen?